src/Razor/src/Microsoft.AspNetCore.Razor.Common/Telemetry/TelemetryReporter.cs

```diff
@@ -73,4 +76,90 @@ private void Report(TelemetryEvent telemetryEvent)
             _logger?.LogError(e, "Failed logging telemetry event");
         }
     }
+
+    public void ReportFault(Exception exception, string? message, object[] @params)
+    {
+        try
+        {
+            if (exception is OperationCanceledException { InnerException: { } oceInnerException })
+            {
+                ReportFault(oceInnerException, message, @params);
+                return;
+            }
+
+            if (exception is AggregateException aggregateException)
+            {
+                // We (potentially) have multiple exceptions; let's just report each of them
+                foreach (var innerException in aggregateException.Flatten().InnerExceptions)
+                    ReportFault(innerException, message, @params);
+
+                return;
+            }
+
+            var currentProcess = Process.GetCurrentProcess();
+
+            var faultEvent = new FaultEvent(
+                eventName: "razor/fault",
+                description: GetDescription(exception),
+                FaultSeverity.General,
+                exceptionObject: exception,
+                gatherEventDetails: faultUtility =>
+                {
+                    // Returning "0" signals that, if sampled, we should send data to Watson. 
+                    // Any other value will cancel the Watson report. We never want to trigger a process dump manually, 
+                    // we'll let TargetedNotifications determine if a dump should be collected.
+                    // See https://aka.ms/roslynnfwdocs for more details
+                    return 0;
```

Comment:
One thing missing here is what the request actually was. It looks like by default this isn't provided to the logger, but maybe we could change that?